pubnub-core/src/data/presence.rs

```diff
@@ -0,0 +1,82 @@
+//! Presence API related types.
+use super::object::Object;
+use super::uuid::UUID;
+use std::collections::HashMap;
+
+pub mod respond_with {
+    //! Type system level flags to specialize the response types.
+
+    /// Return occupancy only.
+    #[derive(Debug, Clone, Copy)]
+    pub struct OccupancyOnly;
+    /// Return occupancy and UUIDs of the users.
+    #[derive(Debug, Clone, Copy)]
+    pub struct OccupancyAndUUIDs;
+    /// Return cooupance, UUIDs of the users and the related states.
+    #[derive(Debug, Clone, Copy)]
+    pub struct Full;
+
+    /// A trait that bounds type system level flag to an actual type
+    /// representing the response structure.
+    pub trait RespondWith {
+        /// The response type that will be produced for the particular
+        /// type system level flag type.
+        type Response;
+    }
+
+    use super::*;
+
+    impl RespondWith for OccupancyOnly {
+        type Response = ChannelInfo;
+    }
+    impl RespondWith for OccupancyAndUUIDs {
+        type Response = ChannelInfoWithOccupants<UUID>;
+    }
+    impl RespondWith for Full {
+        type Response = ChannelInfoWithOccupants<ChannelOccupantFullDetails>;
+    }
```

Comment:
Wondering your thoughts, here... Could this use a struct-like enum instead of traits?

Reply:
The core idea here is to have this handled at the type system level. District types allow us to do this, while having an enum, on the contrary, doesn't allow us to involve the type system. We rely on this to shape the inputs and outputs of the `Transport` service.